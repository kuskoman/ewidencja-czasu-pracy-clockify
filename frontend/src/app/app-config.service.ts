import { Injectable } from '@angular/core';
import { from, lastValueFrom } from 'rxjs';
import { map } from 'rxjs/operators';

@Injectable({ providedIn: 'root' })
export class AppConfigurationService {
  private config$: any = null;

  constructor() {}

  private async load() {
    const defaultConfig$ = from(
      fetch('assets/app-config.default.json').then((response) =>
        response.json()
      )
    );

    const config$ = from(
      fetch('assets/app-config.json').then((response) => response.json())
    );

    const mergedConfig$ = defaultConfig$.pipe(
      map((defaultConfig) => {
        return config$.pipe(
          map((config) => {
            return Object.assign({}, defaultConfig, config);
          })
        );
      })
    );

    this.config$ = await lastValueFrom(mergedConfig$);
  }

  public async get(key: string) {
    if (!this.config$) {
      await this.load();
    }

    return this.config$[key];
  }
}
