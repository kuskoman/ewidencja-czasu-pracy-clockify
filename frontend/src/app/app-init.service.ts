import { Injectable } from '@angular/core';
import { from, lastValueFrom } from 'rxjs';
import { map } from 'rxjs/operators';

declare var window: any;

@Injectable()
export class AppInitService {
  public async init() {
    const defaultConfig$ = from(
      fetch('assets/app-config.default.json').then((response) =>
        response.json()
      )
    );

    const config$ = from(
      fetch('assets/app-config.json').then((response) => response.json())
    );

    const mergedConfig$ = defaultConfig$.pipe(
      map((defaultConfig) => {
        return config$.pipe(
          map((config) => {
            const mergedConfig = Object.assign({}, defaultConfig, config);
            window.config = mergedConfig;
            return;
          })
        );
      })
    );

    return await lastValueFrom(mergedConfig$);
  }
}
