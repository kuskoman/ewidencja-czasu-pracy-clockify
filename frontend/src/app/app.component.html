<!-- A div with main content centered and limited to 800px -->
<div class="main">
  <main class="main__content">
    <app-report-form></app-report-form>
  </main>
</div>
