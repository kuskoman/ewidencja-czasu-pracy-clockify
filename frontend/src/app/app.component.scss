.main__content {
  margin: {
    top: 40px;
    right: auto;
    bottom: 40px;
    left: auto;
  }
  padding: 0 1rem;
  max-width: 800px;
}
