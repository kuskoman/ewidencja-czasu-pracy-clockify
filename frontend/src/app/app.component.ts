import { Component } from '@angular/core';
import { ReportFormComponent } from './report-form/report-form.component';

@Component({
  selector: 'app-root',
  templateUrl: './app.component.html',
  styleUrls: ['./app.component.scss'],
})
export class AppComponent {
  title = 'frontend';
}
