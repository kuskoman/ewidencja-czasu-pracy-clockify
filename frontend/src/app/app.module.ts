import { APP_INITIALIZER, NgModule } from '@angular/core';
import { BrowserModule } from '@angular/platform-browser';

import { AppComponent } from './app.component';
import { BrowserAnimationsModule } from '@angular/platform-browser/animations';
import { ReportFormComponent } from './report-form/report-form.component';
import { NavbarComponent } from './navbar/navbar.component';
import { FooterComponent } from './footer/footer.component';
import { AppInitService } from './app-init.service';

export const initApp = (appInitService: AppInitService) => {
  return () => appInitService.init();
};

@NgModule({
  declarations: [AppComponent],
  imports: [
    BrowserModule,
    BrowserAnimationsModule,
    ReportFormComponent,
    NavbarComponent,
    FooterComponent,
  ],
  providers: [
    AppInitService,
    {
      provide: APP_INITIALIZER,
      useFactory: initApp,
      deps: [AppInitService],
      multi: true,
    },
  ],
  bootstrap: [AppComponent],
})
export class AppModule {}
