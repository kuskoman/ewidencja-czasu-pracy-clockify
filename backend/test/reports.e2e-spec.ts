import { Test, TestingModule } from '@nestjs/testing';
import { INestApplication, ValidationPipe } from '@nestjs/common';
import * as request from 'supertest';
import { AppModule } from '../src/app.module';
import { join } from 'path';

describe('ReportsController (e2e)', () => {
  let app: INestApplication;

  beforeEach(async () => {
    const moduleFixture: TestingModule = await Test.createTestingModule({
      imports: [AppModule],
    }).compile();

    app = moduleFixture.createNestApplication();
    app.useGlobalPipes(
      new ValidationPipe({
        whitelist: true,
        errorHttpStatusCode: 422,
        transform: true,
      }),
    );
    app.enableShutdownHooks();
    await app.init();
  });

  afterEach(async () => {
    await app.close();
  });

  describe('port /reports', () => {
    it('should return 200 and create a report when valid CSV file and data is provided', async () => {
      const response = await request(app.getHttpServer())
        .post('/reports')
        .attach('file', getFilePath('validReport.csv'))
        .field('name', 'Jakub')
        .field('surname', 'Surdej');

      expect(response.status).toBe(200);
      expect(response.body).toBeDefined();
      expect(response.type).toBe('text/html');
      expect(response.text).toMatchSnapshot();
    });

    it('should return 422 when invalid data is provided', async () => {
      const response = await request(app.getHttpServer())
        .post('/reports')
        .attach('file', getFilePath('invalidReport.csv'))
        .field('name', 'Jakub')
        .field('surname', 'Surdej');

      expect(response.status).toBe(422);
      expect(response.body).toBeDefined();
      // todo: check if the error message is correct
    });

    it('should return 422 when no CSV file is provided', async () => {
      const response = await request(app.getHttpServer())
        .post('/reports')
        .field('name', 'Jakub')
        .field('surname', 'Surdej');

      expect(response.status).toBe(422);
      expect(response.body).toBeDefined();
      expect(response.body).toHaveProperty(
        'message',
        'You must provide a valid file',
      );
    });
  });
});

const getFilePath = (fileName: string) => join(__dirname, 'fixtures', fileName);
