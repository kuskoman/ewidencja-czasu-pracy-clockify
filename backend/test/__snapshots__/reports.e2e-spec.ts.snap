// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReportsController (e2e) port /reports should return 200 and create a report when valid CSV file and data is provided 1`] = `
"<!DOCTYPE html>
<html>
  <head>
    <title>Ewidencja Czasu Pracy</title>
  </head>
  <body>
    <main>
      <h1>EWIDENCJA CZASU ZLECENIOBIORCY</h1>
      <p><span style="font-weight: bold">Okres:</span> 2023/02/01 - 2023/02/28</p>
      <p><span style="font-weight: bold">Imię:</span> Jakub</p>
      <p>
        <span style="font-weight: bold">Nazwisko:</span> Surdej
      </p>

      <table>
        <tr>
          <td><span style="font-weight: bold">Data</span></td>
          <td><span style="font-weight: bold">od</span></td>
          <td><span style="font-weight: bold">do</span></td>
          <td><span style="font-weight: bold">ilość godzin</span></td>
          <td><span style="font-weight: bold">podpis pracownika</span></td>
        </tr>

        
        <tr>
          <td>2023/02/01</td>
          <td>8:30</td>
          <td>17:30</td>
          <td>9</td>
          <td></td>
        </tr>
        
        <tr>
          <td>2023/02/02</td>
          <td>8:30</td>
          <td>17:30</td>
          <td>9</td>
          <td></td>
        </tr>
        
        <tr>
          <td>2023/02/03</td>
          <td>8:30</td>
          <td>15:30</td>
          <td>7</td>
          <td></td>
        </tr>
        
        <tr>
          <td>2023/02/06</td>
          <td>8:30</td>
          <td>16:30</td>
          <td>8</td>
          <td></td>
        </tr>
        
        <tr>
          <td>2023/02/07</td>
          <td>8:30</td>
          <td>16:30</td>
          <td>8</td>
          <td></td>
        </tr>
        
        <tr>
          <td>2023/02/08</td>
          <td>8:30</td>
          <td>16:30</td>
          <td>8</td>
          <td></td>
        </tr>
        
        <tr>
          <td>2023/02/09</td>
          <td>8:30</td>
          <td>16:30</td>
          <td>8</td>
          <td></td>
        </tr>
        
        <tr>
          <td>2023/02/10</td>
          <td>8:30</td>
          <td>16:30</td>
          <td>8</td>
          <td></td>
        </tr>
        
        <tr>
          <td>2023/02/13</td>
          <td>8:30</td>
          <td>18:30</td>
          <td>10</td>
          <td></td>
        </tr>
        
        <tr>
          <td>2023/02/14</td>
          <td>8:30</td>
          <td>14:30</td>
          <td>6</td>
          <td></td>
        </tr>
        
        <tr>
          <td>2023/02/15</td>
          <td>8:30</td>
          <td>16:30</td>
          <td>8</td>
          <td></td>
        </tr>
        
        <tr>
          <td>2023/02/16</td>
          <td>8:30</td>
          <td>16:30</td>
          <td>8</td>
          <td></td>
        </tr>
        
        <tr>
          <td>2023/02/17</td>
          <td>8:30</td>
          <td>16:30</td>
          <td>8</td>
          <td></td>
        </tr>
        
        <tr>
          <td>2023/02/20</td>
          <td>8:30</td>
          <td>16:30</td>
          <td>8</td>
          <td></td>
        </tr>
        
        <tr>
          <td>2023/02/21</td>
          <td>8:30</td>
          <td>16:30</td>
          <td>8</td>
          <td></td>
        </tr>
        
        <tr>
          <td>2023/02/22</td>
          <td>8:30</td>
          <td>16:30</td>
          <td>8</td>
          <td></td>
        </tr>
        
        <tr>
          <td>2023/02/23</td>
          <td>8:30</td>
          <td>16:30</td>
          <td>8</td>
          <td></td>
        </tr>
        
        <tr>
          <td>2023/02/24</td>
          <td>8:30</td>
          <td>16:30</td>
          <td>8</td>
          <td></td>
        </tr>
        
        <tr>
          <td>2023/02/27</td>
          <td>8:30</td>
          <td>16:30</td>
          <td>8</td>
          <td></td>
        </tr>
        
        <tr>
          <td>2023/02/28</td>
          <td>8:30</td>
          <td>16:30</td>
          <td>8</td>
          <td></td>
        </tr>
        

        <tr>
          <td><span style="font-weight: bold">Suma</span></td>
          <td></td>
          <td></td>
          <td><span style="font-weight: bold">160</span></td>
          <td></td>
        </tr>
      </table>
    </main>
    <style>
      @page {
        size: A4;
      }

      @page {
        margin-left: 5px;
        margin-right: 5px;
      }

      main {
        text-align: center;
      }

      table {
        margin: auto;
      }

      table,
      td {
        border: 1px solid black;
        border-collapse: collapse;
        text-align: center;
      }

      td {
        min-width: 75px;
      }
    </style>
  </body>
</html>
"
`;
